src/ripple/app/ledger/impl/DisputedTx.h

```diff
@@ -73,26 +77,158 @@ class DisputedTx
         mOurVote = o;
     }
 
-    // VFALCO NOTE its not really a peer, its the 160 bit hash of the
-    // validator's public key.
-    void setVote (NodeID const& peer, bool votesYes);
-    void unVote (NodeID const& peer);
+    void setVote (NodeID_t const& peer, bool votesYes);
+    void unVote (NodeID_t const& peer);
 
     bool updateVote (int percentTime, bool proposing);
     Json::Value getJson ();
 
 private:
-    uint256 mTransactionID;
+    TxID_t mTransactionID;
     int mYays;
     int mNays;
     bool mOurVote;
-    // VFALCO Why is this being stored as a Serializer?
-    Serializer transaction;
+    Tx_t transaction;
 
-    hash_map <NodeID, bool> mVotes;
+    hash_map <NodeID_t, bool> mVotes;
     beast::Journal j_;
 };
 
+// Track a peer's yes/no vote on a particular disputed transaction
+template <class Traits>
+void DisputedTx<Traits>::setVote (NodeID_t const& peer, bool votesYes)
+{
+    auto res = mVotes.insert (std::make_pair (peer, votesYes));
+
+    // new vote
+    if (res.second)
+    {
+        if (votesYes)
+        {
+            JLOG (j_.debug())
+                    << "Peer " << peer << " votes YES on " << mTransactionID;
+            ++mYays;
+        }
+        else
+        {
+            JLOG (j_.debug())
+                    << "Peer " << peer << " votes NO on " << mTransactionID;
+            ++mNays;
+        }
+    }
+    // changes vote to yes
+    else if (votesYes && !res.first->second)
+    {
+        JLOG (j_.debug())
+                << "Peer " << peer << " now votes YES on " << mTransactionID;
+        --mNays;
+        ++mYays;
+        res.first->second = true;
+    }
+    // changes vote to no
+    else if (!votesYes && res.first->second)
+    {
+        JLOG (j_.debug())
+                << "Peer " << peer << " now votes NO on " << mTransactionID;
+        ++mNays;
+        --mYays;
+        res.first->second = false;
+    }
+}
+
+// Remove a peer's vote on this disputed transasction
+template <class Traits>
+void DisputedTx<Traits>::unVote (NodeID_t const& peer)
+{
+    auto it = mVotes.find (peer);
+
+    if (it != mVotes.end ())
+    {
+        if (it->second)
+            --mYays;
+        else
+            --mNays;
+
+        mVotes.erase (it);
+    }
+}
+
+template <class Traits>
+bool DisputedTx<Traits>::updateVote (int percentTime, bool proposing)
+{
+    if (mOurVote && (mNays == 0))
+        return false;
+
+    if (!mOurVote && (mYays == 0))
+        return false;
+
+    bool newPosition;
+    int weight;
+
+    if (proposing) // give ourselves full weight
+    {
+        // This is basically the percentage of nodes voting 'yes' (including us)
+        weight = (mYays * 100 + (mOurVote ? 100 : 0)) / (mNays + mYays + 1);
```

Comment:
I'm not sure if this is simpler or not, but it's an option and avoids the ternary:

    int weight = ((mYays + static_cast<int>(mOurVote)) * 100) / (mNays + mYays + 1);

This is guaranteed safe and correct by §4.7/4 of the C++ Standard, discussing integral conversions:

> If the source type is bool, the value false is converted to zero and the value true is converted to one.